### ISSUE TITLE
Perl::Critic::Policy's violates method print to stderr without throwing an exception

### ISSUE THREAD

**Issue by edenhochbaum** [Perl-Critic/Perl-Critic#657]
Is there guidance on when a Perl::Critic::Policy's violates method should print to stderr *without* throwing an exception?  For example, Perl::Critic::Policy::Modules::RequireExplicitInclusion (not part of this repot, but rather part of https://github.com/Perl-Critic/Perl-Critic-StricterSubs) gives a warning like:

```perl
   warn qq{$pname: Cannot cope with mutiple packages in file "$fname"\n};
```
before just returning.  Most policies just return when they can't cope, though.

**Comment by thaljef** [comment]
Good question. In general, I think Policies should only throw an exception if there's a failure in Perl::Critic itself, never because of the input code.

Some Policies use the `prepare_to_scan_document` method to decide if the Policy should run at all for a given file. For example, if the Policy only applies to modules, but the input file is a script, or if the Policy only applies to code that is targeting a certain version of Perl.

You could argue that `RequireExplicitInclusion` should also use the `prepare_to_scan_document` mechanism instead of issuing a warning. I think the Policy may have been written before it existed.

The whole StricterSubs distribution was written for a particular client who was trying to prevent specific patterns in their code, which probably included multiple packages in the same file. So for them, the warning was unoffensive because they also used the `ProhibitMultiplePackage` Policy. But in general, each Policy should behave sensibly regardless of the other Policies that are being used.

So basically, it depends. Shoot for DWIM. When in doubt, document the behavior.

Hope that helps.

**Comment by edenhochbaum** [comment]
Thanks!  I've just commented this out for my own use, am closing the issue.

**Comment by thaljef** [comment]
> I've just commented this out for my own use, am closing the issue.

I see. If you usually want to follow the `RequireExplicitInclusion` Policy, but you have a few files that use multiple packages, then you could just disable the Policy in those files with a `## no critic` annotation. That should prevent the warning and it also expresses your intent more clearly. Does that make sense?

**Comment by edenhochbaum** [comment]
that makes total sense, definitely - thanks for your help!
